### ISSUE TITLE
Dolphin Emulator - Alpha Test and Discard failure with "UberShaders"

### ISSUE THREAD

**Issue by Sonicadvance1** [daniel-schuermann/mesa#67]
Looks like when Ubershaders are enabled with ACO then when Alpha testing and doing a discard, the depth is still written to the depth buffer resulting in rendering artifacts. Potentially similar to #65.
There also looks to be some graphical corruption where the alpha test path is failing in their shader.
Or potentially UV calculation getting broken, since the texture corruption looks like a small portion of the texture is supposed to get through but it ends up looking like corruption with ACO.

This doesn't suffer from derivative calculations after Kill. All texturing is done prior to the kill.
Additionally when dropping to their specialized shaders, the corruption goes away.
Of course dropping back to LLVM for either path fixes it.

Seeing if I can fix it myself but I can't guarantee it.

Corruption from draw call
![AC-UberCorruption](https://user-images.githubusercontent.com/1018829/60854399-0239b680-a1b5-11e9-91d1-89dab7db4e74.png)

Non-corruption from LLVM
![AC-UberLLVM](https://user-images.githubusercontent.com/1018829/60854415-0f56a580-a1b5-11e9-8dd2-2cf0ee3f5762.png)

In the renderdoc capture it is event ID 155 showing the corruption here.
[RenderDocCapture.zip](https://github.com/daniel-schuermann/mesa/files/3370900/RenderDocCapture.zip)

**Comment by Sonicadvance1** [comment]
Nested conditional discard is causing the rendering artifacts. Change to Dolphin below works around the issue.
Looking in to an ACO fix for it.


```--- a/Source/Core/VideoCommon/UberShaderPixel.cpp
+++ b/Source/Core/VideoCommon/UberShaderPixel.cpp
@@ -1119,6 +1119,7 @@ ShaderCode GenPixelShader(APIType ApiType, const ShaderHostConfig& host_config,
       out.Write("  depth = float(zbuffer_zCoord) / 16777216.0;\n");
   }

+  out.Write("bool NeedsToDie = false;\n");
   out.Write("  // Alpha Test\n"
             "  if (bpmem_alphaTest != 0u) {\n"
             "    bool comp0 = alphaCompare(TevResult.a, " I_ALPHA ".r, %s);\n",
@@ -1131,16 +1132,18 @@ ShaderCode GenPixelShader(APIType ApiType, const ShaderHostConfig& host_config,
             "    switch (%s) {\n",
             BitfieldExtract("bpmem_alphaTest", AlphaTest().logic).c_str());
   out.Write("    case 0u: // AND\n"
-            "      if (comp0 && comp1) break; else discard; break;\n"
+            "      if (comp0 && comp1) break; else NeedsToDie = true; break;\n"
             "    case 1u: // OR\n"
-            "      if (comp0 || comp1) break; else discard; break;\n"
+            "      if (comp0 || comp1) break; else NeedsToDie = true; break;\n"
             "    case 2u: // XOR\n"
-            "      if (comp0 != comp1) break; else discard; break;\n"
+            "      if (comp0 != comp1) break; else NeedsToDie = true; break;\n"
             "    case 3u: // XNOR\n"
-            "      if (comp0 == comp1) break; else discard; break;\n"
+            "      if (comp0 == comp1) break; else NeedsToDie = true; break;\n"
+               "    default: break;\n"
             "    }\n"
             "  }\n"
             "\n");
+  out.Write("if (NeedsToDie) { discard; } \n");

   // =========
   // Dithering```

**Comment by Sonicadvance1** [comment]
Shader code past the GLSL->SPIR-V and renderdoc path prior to working around the problem

```  if(PSBlock_244.bpmem_alphaTest != 0) {
    BitFieldUExtract(_1974_, 16, 3);
    comp0 = alphaCompare(TevResult.w, PSBlock_244.alphaRef.x, BitFieldUExtract(_1974_, 16, 3));
    BitFieldUExtract(_1986_, 19, 3);
    comp1 = alphaCompare(TevResult.w, PSBlock_244.alphaRef.y, BitFieldUExtract(_1986_, 19, 3));
    switch(BitFieldUExtract(_1997_, 22, 2)) {
    case 0:
      if(comp0 && comp1) {
        break;
      } else {
        Kill();
      }
      break;
    case 1:
      if(comp0 || comp1) {
        break;
      } else {
        Kill();
      }
      break;
    case 2:
      if(comp0 != comp1) {
        break;
      } else {
        Kill();
      }
      break;
    case 3:
      if(comp0 == comp1) {
        break;
      } else {
        Kill();
      }
      break;
    }
  }```

**Comment by daniel-schuermann** [comment]
fixed with 2064386e7340d5db9f06aa4e5e9ea2639c267e24 (master) and 8c6ebf61d7ab4454b06e195664de631948c3ec1b (backend)
